janus_core/log.py

```diff
@@ -5,13 +5,83 @@
 
 from janus_core.janus_types import LogLevel
 
-FORMAT = """
-- timestamp: %(asctime)s
-  level: %(levelname)s
-  message: %(message)s
-  trace: %(module)s
-  line: %(lineno)d
-""".strip()
+
+class YamlFormatter(logging.Formatter):  # numpydoc ignore=PR02
+    """
+    Custom formatter to convert multiline messages into yaml list.
+
+    Parameters
+    ----------
+    fmt : Optional[str]
+        A format string in the given style for the logged output as a whole. Default is
+        defined by `FORMAT`.
+    datefmt : Optional[str]
+        A format string in the given style for the date/time portion of the logged
+        output. Default is taken from logging.Formatter.formatTime().
+    style : Literal['%', '{', '$']
+        Determines how the format string will be merged with its data. Can be one of
+        '%', '{' or '$'. Default is '%'.
+    validate : bool
+        If True, incorrect or mismatched fmt and style will raise a ValueError. Default
+        is True.
+    defaults : Optional[dict[str, logging.Any]]
+         A dictionary with default values to use in custom fields. Default is None.
+    *args
+        Arguments to pass to logging.Formatter.__init__.
+    **kwargs
+        Keyword arguments to pass to logging.Formatter.__init__.
+
+    Methods
+    -------
+    format(record)
+        Format log message to convert new lines into a yaml list.
+    """
+
+    FORMAT = (
+        "\n- timestamp: %(asctime)s\n"
+        "  level: %(levelname)s\n"
+        "  message: %(message)s\n"
+        "  trace: %(module)s\n"
+        "  line: %(lineno)d\n"
+    )
+
+    def __init__(self, *args, **kwargs) -> None:
+        """
+        Set default string format to yaml style.
+
+        Parameters
+        ----------
+        *args
+            Arguments to pass to logging.Formatter.__init__.
+        **kwargs
+            Keyword arguments to pass to logging.Formatter.__init__.
+        """
+        kwargs.setdefault("fmt", self.FORMAT)
+        super().__init__(*args, **kwargs)
+
+    def format(self, record: logging.LogRecord) -> str:
+        """
+        Format log message to convert new lines into a yaml list.
+
+        Parameters
+        ----------
+        record : logging.LogRecord
+            Log record to be formatted.
+
+        Returns
+        -------
+        str
+            Formatted log message.
+        """
+        # Replace "" with '' to prevent invalid wrapping
+        record.msg = record.msg.replace('"', "'")
+
+        # Convert new lines into yaml list
+        msg = record.msg
+        record.msg = "\n" + "\n".join(
+            f'    - "{line.strip()}"' for line in msg.split("\n") if line.strip()
+        )
```

Comment:
Is `msg` necessary anymore now this is a comprehension?
```suggestion
        record.msg = "\n" + "\n".join(
            f'    - "{line.strip()}"' for line in record.msg.splitlines() if line.strip()
        )
```

Reply:
No, good point